src/renderer/base/RenderSettings.cpp

```diff
@@ -0,0 +1,289 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "precomp.h"
+
+#include "../inc/RenderSettings.hpp"
+#include "../inc/IRenderTarget.hpp"
+#include "../../types/inc/ColorFix.hpp"
+#include "../../types/inc/colorTable.hpp"
+
+using namespace Microsoft::Console::Render;
+using Microsoft::Console::Utils::InitializeColorTable;
+
+static constexpr size_t AdjustedFgIndex{ 16 };
+static constexpr size_t AdjustedBgIndex{ 17 };
+
+RenderSettings::RenderSettings() noexcept
+{
+    InitializeColorTable(_colorTable);
+
+    SetColorTableEntry(TextColor::DEFAULT_FOREGROUND, INVALID_COLOR);
+    SetColorTableEntry(TextColor::DEFAULT_BACKGROUND, INVALID_COLOR);
+    SetColorTableEntry(TextColor::CURSOR_COLOR, INVALID_COLOR);
+
+    SetColorAliasIndex(ColorAlias::DefaultForeground, TextColor::DARK_WHITE);
+    SetColorAliasIndex(ColorAlias::DefaultBackground, TextColor::DARK_BLACK);
+}
+
+// Routine Description:
+// - Updates the specified render mode.
+// Arguments:
+// - mode - The render mode to change.
+// - enabled - Set to true to enable the mode, false to disable it.
+void RenderSettings::SetRenderMode(const Mode mode, const bool enabled) noexcept
+{
+    _renderMode.set(mode, enabled);
+    // If blinking is disabled, make sure blinking content is not faint.
+    if (mode == Mode::BlinkAllowed && !enabled)
+    {
+        _blinkShouldBeFaint = false;
+    }
+}
+
+// Routine Description:
+// - Retrieves the specified render mode.
+// Arguments:
+// - mode - The render mode to query.
+// Return Value:
+// - True if the mode is enabled. False if disabled.
+bool RenderSettings::GetRenderMode(const Mode mode) const noexcept
+{
+    return _renderMode.test(mode);
+}
+
+// Routine Description:
+// - Returns a reference to the active color table array.
+const std::array<COLORREF, TextColor::TABLE_SIZE>& RenderSettings::GetColorTable() const noexcept
+{
+    return _colorTable;
+}
+
+// Routine Description:
+// - Resets the first 16 color table entries with default values.
+void RenderSettings::ResetColorTable() noexcept
+{
+    InitializeColorTable({ _colorTable.data(), 16 });
+}
+
+// Routine Description:
+// - Creates the adjusted color array, which contains the possible foreground colors,
+//   adjusted for perceivability
+// - The adjusted color array is 2-d, and effectively maps a background and foreground
+//   color pair to the adjusted foreground for that color pair
+void RenderSettings::MakeAdjustedColorArray() noexcept
+{
+    // The color table has 16 colors, but the adjusted color table needs to be 18

```

Comment:
FWIW I believe that we now calculate this for everyone, even if the feature isn't being used -- that might have a cost in conhost that we don't want to take on, right?

Reply:
No, right now it's only being called from Terminal.cpp ([here](https://github.com/microsoft/terminal/blob/62c95b5017e92a780cdc43008e30b4e43d2edc9b/src/cascadia/TerminalCore/Terminal.cpp#L190) and [here](https://github.com/microsoft/terminal/blob/62c95b5017e92a780cdc43008e30b4e43d2edc9b/src/cascadia/TerminalCore/Terminal.cpp#L1331)).

There is some slight overhead in the `GetAttributeColors` method testing the `DistinguishableColors` mode (see [here](https://github.com/microsoft/terminal/blob/62c95b5017e92a780cdc43008e30b4e43d2edc9b/src/renderer/base/RenderSettings.cpp#L189)), which won't be set in conhost, but I'm still hopeful we can get rid of that whole branch at some point.

Reply:
I've tested this PR locally before approving it. I found no regressions. Binary size (without PGO) only increased ever so slightly (~500 Byte).